From de3d2bbdf5c08d20b84b730e5ee0c7fcdb73a972 Mon Sep 17 00:00:00 2001
From: Sami Farin <hvtaifwkbgefbaei@gmail.com>
Date: Fri, 17 Jan 2025 11:51:50 +0200
Subject: [PATCH] write window widthXheight to config file

something still changes the size on each start by 99 pixels or so
and config file writes can happen hundred times a second :D (fix coming)
---
 src/spek-preferences.cc | 18 ++++++++++++++++++
 src/spek-preferences.h  |  7 +++++++
 src/spek-spectrogram.cc |  5 +++++
 src/spek-window.cc      |  8 +++++++-
 4 files changed, 37 insertions(+), 1 deletion(-)

diff --git a/src/spek-preferences.cc b/src/spek-preferences.cc
index 67cbd098..0650d338 100644
--- a/src/spek-preferences.cc
+++ b/src/spek-preferences.cc
@@ -1,4 +1,6 @@
 #include <wx/string.h>
+#include <wx/window.h>
+#include <wx/dcbuffer.h>
 
 #include "spek-platform.h"
 
@@ -49,6 +51,22 @@ bool SpekPreferences::get_check_update()
     return result;
 }
 
+wxSize SpekPreferences::get_window_size()
+{
+    int width = SpekPreferences::DEF_WIDTH;
+    int height = SpekPreferences::DEF_HEIGHT;
+    this->config->Read("/window/width", &width);
+    this->config->Read("/window/height", &height);
+    return wxSize(width, height);
+}
+
+void SpekPreferences::set_window_size(int width, int height)
+{
+    this->config->Write("/window/width", width);
+    this->config->Write("/window/height", height);
+    this->config->Flush();
+}
+
 void SpekPreferences::set_check_update(bool value)
 {
     this->config->Write("/update/check", value);
diff --git a/src/spek-preferences.h b/src/spek-preferences.h
index 3ccc5ca8..733f65b2 100644
--- a/src/spek-preferences.h
+++ b/src/spek-preferences.h
@@ -2,6 +2,8 @@
 
 #include <wx/fileconf.h>
 #include <wx/intl.h>
+#include <wx/window.h>
+#include <wx/dcbuffer.h>
 
 class SpekPreferences
 {
@@ -15,6 +17,11 @@ class SpekPreferences
     void set_last_update(long value);
     wxString get_language();
     void set_language(const wxString& value);
+    void set_window_size(int width, int height);
+    wxSize get_window_size();
+
+    static const int DEF_WIDTH = 800;
+    static const int DEF_HEIGHT = 600;
 
 private:
     SpekPreferences();
diff --git a/src/spek-spectrogram.cc b/src/spek-spectrogram.cc
index cc8d4780..990d2cbd 100644
--- a/src/spek-spectrogram.cc
+++ b/src/spek-spectrogram.cc
@@ -8,6 +8,7 @@
 #include "spek-platform.h"
 #include "spek-ruler.h"
 #include "spek-utils.h"
+#include "spek-preferences.h"
 
 #include "spek-spectrogram.h"
 
@@ -167,9 +168,13 @@ void SpekSpectrogram::on_paint(wxPaintEvent&)
 void SpekSpectrogram::on_size(wxSizeEvent&)
 {
     wxSize size = GetClientSize();
+    wxLogDebug("wxSize [on_size] = %dx%d", size.GetWidth(), size.GetHeight());
     bool width_changed = this->prev_width != size.GetWidth();
     this->prev_width = size.GetWidth();
 
+    // Save the new window size
+    SpekPreferences::get().set_window_size(size.GetWidth(), size.GetHeight());
+
     if (width_changed) {
         start();
     }
diff --git a/src/spek-window.cc b/src/spek-window.cc
index 86d360cf..d3c2524b 100644
--- a/src/spek-window.cc
+++ b/src/spek-window.cc
@@ -56,7 +56,13 @@ SpekWindow::SpekWindow(const wxString& path) :
 {
     this->description = _("Spek - Acoustic Spectrum Analyser");
     SetTitle(this->description);
-    SetSize(this->FromDIP(wxSize(640, 480)));
+    wxSize size = SpekPreferences::get().get_window_size();
+    wxLogDebug("wxSize [prefs] = %dx%d", size.GetWidth(), size.GetHeight());
+    wxSize defsz = wxSize(SpekPreferences::DEF_WIDTH, SpekPreferences::DEF_HEIGHT);
+    if (!size.GetWidth() || !size.GetHeight()) size = defsz;
+    wxSize newsz = this->FromDIP(size);
+    wxLogDebug("wxSize [new] =  %dx%d", newsz.GetWidth(), newsz.GetHeight());
+    SetSize(newsz);
 
 #ifndef OS_OSX
     SetIcons(wxArtProvider::GetIconBundle(ART_SPEK, wxART_FRAME_ICON));
